### ISSUE TITLE
Can I use third-party components?

### ISSUE THREAD

**Issue by yuanotes** [webcodesk/webcodesk#4]
Can I use components like Antd https://ant.design/docs/react/introduce ?
It would make a huge difference.

**Comment by ipselon** [comment]
@yuanotes, sure.

You have to place your components into `src/usr` folder. And your components should have the correct syntax.

There are two ways to declare components in the code by which Webcodesk will recognize them as the React components and will include them into the available components list.

* A **functional component** with the name equal to the file name, and with the default export in the file (thus, one file - one component)
* A **class declaration** with the name equal to the file name, and with the default export (one file - one component)

Add properties with **PropTypes.element** type to the component declaration if you want to combine components in the page editor.

Declare "output" properties (events) with **PropTypes.func** type.

Finally, add properties with any **other types** to make a component with inputs on the diagram.

Use the following as an example:

```
import React from 'react';
import PropTypes from 'prop-types';

const rootStyle = {
  width: '150px',
  height: '150px',
  display: 'flex',
  flexDirection: 'column',
  justifyContent: 'center',
  alignItems: 'center',
};

const inputStyle = {
  padding: '5px',
  borderRadius: '4px',
  border: '1px solid #cccccc',
};

const buttonStyle = {
  padding: '5px',
  borderRadius: '4px',
};

/*
  Input form for the user name
 */
class Form extends React.Component {
  static propTypes = {
    // send the entered name
    onClick: PropTypes.func,
  };
  
  static defaultProps = {
    onClick: () => {
      // is not set
    },
  };

  handleClick = (e) => {
    if (e) {
      e.stopPropagation();
      e.preventDefault();
    }
    this.props.onClick(this.inputElement.value);
  };

  render () {
    return (
      <form onSubmit={this.handleClick}>
        <div style={rootStyle}>
          <div style={{ margin: '1em 0 1em 0' }}>
            <input
              ref={me => this.inputElement = me}
              type="text"
              placeholder="Enter your name"
              style={inputStyle}
            />
          </div>
          <div>
            <button
              type="submit"
              onClick={this.handleClick}
              style={buttonStyle}
            >
              Click
            </button>
          </div>
        </div>
      </form>
    );
  }
}

export default Form;
```

**Comment by yuanotes** [comment]
Thanks for the reply. 

Actually, if I can use UI libraries like material ui or ant design **directly** by adding them to dependencies in package.json,  it would be quite helpful. 

It means that I can build a production ready webapp in a few minutes.

And you will have a lot of paid subscriptions. :)

**Comment by ipselon** [comment]
@yuanotes,

That's possible to add any dependency into the project. However, you have to add the component code which includes the library components. Webcodesk does not generate the code for components itself.

For example, take a look (download and open in Webcodesk) at this project [firebase_auth_tutorial](https://github.com/ipselon/firebase_auth_tutorial) - it uses Material-UI library.

**Comment by ipselon** [comment]
Hey @yuanotes. 

I bumped a new version with the ability to install React components from the marketplace. But there is also the ability to publish components to the market for anyone.

**Comment by yuanotes** [comment]
Thanks Alex. The idea of marketplace is great.

It seems that webcodesk detects user defined component by file name and the
proptypes declaration in that file.

Why not parsing the code in node_modules directly instead of make adapters
for them?

For example, the code below would just export all material components to
webcodesk with proptypes.

in user/awesome/index.js

export * from '@material-ui/core';



Alex Pustovalov <notifications@github.com>于2019年4月15日 周一17:07写道：

**Comment by ipselon** [comment]
Yeah, @yuanotes, I see what you mean. However, Webcodesk should understand what the properties component has - and unfortunately there is no way to parse the source code inside the npm module.

**Comment by ipselon** [comment]
@yuanotes, from my practice, working with components without logic inside does not make sense in Webcodesk. And all Material-UI components is too low level - they should be composed into bigger components to be useful in the project. But, of course, these components should be reusable... I guess.

**Comment by yuanotes** [comment]
I see it. Maybe someday I can download or buy components from the online
market.

Have you post the app to Hackernews or Producthunt? It may help get a lot
of attentions.

Alex Pustovalov <notifications@github.com>于2019年4月15日 周一17:53写道：

**Comment by ipselon** [comment]
@yuanotes, hey, I'm looking for the partner - I'm working alone, and have many plans so far. But it's quite hard to make all things all together in parallel. If you can help with the promotion of Webcodesk, I'd much appreciate it.

As for the components, I think about the implementation of the publishing and sharing private components through the market. Then customers may ask developers to contribute needed components to the market in private projects.


However, if you need some complex React components, I can create them for you - I'm a freelancer and would be glad to work with you.

**Comment by yuanotes** [comment]
I’ve had the idea to build a web app builder just like Unity3D. In my
opinion, Webcodesk has a great potential.

Though I have a full time job, I would like to help to make it a success.

Maybe you can make me a referral url or something like that. ;)

Alex Pustovalov <notifications@github.com>于2019年4月15日 周一18:07写道：

**Comment by ipselon** [comment]
@yuanotes. Yeah, that's a great idea. However, there is a significant problem for such builders - developers do not like visual builders. The builder should be perfect for doing something that devs do not like to do manually.

There is great potential for a tool that makes designers and devs work consistently together. But creating such a tool requires hard work and money.

As for the referral URL, I'm not sure that clearly understand what this is... Could you elaborate on this?

**Comment by yuanotes** [comment]
The referral link is a method for promotion. In China, almost every app
tries to reward users for inviting friends to join.

It’s a url with a unique code to the downloading page of an app. It will
tracks how many new downloads are associated with the code.

I can help you to promote webcodesk on forums and social media platforms
and even in large IT companies.

But now it still needs improvements to make it more attractive.

More usable components are needed. Users can make a functional web app with
a few clicks. That will be awesome.



Alex Pustovalov <notifications@github.com>于2019年4月15日 周一18:32写道：

**Comment by ipselon** [comment]
@yuanotes. Aha, I see... 

> I can help you to promote webcodesk on forums and social media platforms
and even in large IT companies.

That would be awesome. When you download Webcodesk, it has no license and works in limited mode - limit: 5 flows and 5 pages. But everyone who creates an account on Webcodesk site gets 60 days of free license - unlimited flows and pages. Is it enough for promoting or should promotion code?

> More usable components are needed. Users can make a functional web app with
a few clicks. That will be awesome.

That's fairly true. I'm thinking now what components should be there to attract companies to start using Webcodesk. I thought about some financial components that can be connected in an application. 

Do you have any examples?

**Comment by ipselon** [comment]
@yuanotes would you mind if we switch to Slack for chatting? I can send you an invite to the group in Slack - give your email if you don't mind.

**Comment by yuanotes** [comment]
That would be great. My email address is  yuanotes@gmail.com.

Alex Pustovalov <notifications@github.com>于2019年4月15日 周一19:38写道：

**Comment by ipselon** [comment]
@yuanotes, done.

**Comment by henrykempo** [comment]
My 2cents worth about this - better not to try to monetise early. Suggest open source route, Dev's can get paid by contributions - gitcoin one way, Later offer premium options - like enterprise edition /support, bespoke consultancy work. No one likes feature throttling - will loose a lot of potential interest via that route. imho

**Comment by ipselon** [comment]
@henrykempo, thank you.

I agree with you. The beta version was private just for the beta testing period. And now I'm preparing all to open source Webcodesk.

**Comment by henrykempo** [comment]
Thats great. Glad you were thinking of the same route. Just saw that you have Open Sourced It too - well 47 mins ago! Nice One Alex! Onwards and Upwards!

**Comment by mdocter** [comment]
> Yeah, @yuanotes, I see what you mean. However, Webcodesk should understand what the properties component has - and unfortunately there is no way to parse the source code inside the npm module.

Can you explain why it wouldn't be possible to `parse the source code inside the npm module`? Does electron prevent this?

**Comment by ipselon** [comment]
@mdocter, it might be I used the wrong word. I tried to explain that way not every library author includes the source code into the package.

Even though there is the source code, we can not guarantee that the sources added to the package are the same code which was used for compilation.

However, I'm planning to add a component design feature, where the user can design the component look and feel just like the other designing tools, and then generate the code for the complex component. I believe that in this case Webcodesk should somehow parse and find components in `node_modules`. I'm not sure how to do this now, but definitely, this is possible from the technical point of view.
